packages/espressocash_app/lib/features/activities/widgets/recent_token_activity.dart

```diff
@@ -0,0 +1,218 @@
+import 'package:fast_immutable_collections/fast_immutable_collections.dart';
+import 'package:flutter/material.dart';
+import 'package:intl/intl.dart';
+
+import '../../../di.dart';
+import '../../../l10n/l10n.dart';
+import '../../../ui/colors.dart';
+import '../../../ui/home_tile.dart';
+import '../../../ui/theme.dart';
+import '../data/transaction_repository.dart';
+import '../models/transaction.dart';
+import 'common_tile.dart';
+import 'odp_tile.dart';
+import 'off_ramp_tile.dart';
+import 'olp_tile.dart';
+import 'on_ramp_tile.dart';
+import 'outgoing_dln_tile.dart';
+import 'payment_request_tile.dart';
+import 'tr_tile.dart';
+
+class RecentTokenActivityWidget extends StatefulWidget {
+  const RecentTokenActivityWidget({
+    super.key,
+    required this.tokenAddress,
+  });
+
+  final String tokenAddress;
+
+  @override
+  State<RecentTokenActivityWidget> createState() =>
+      _RecentTokenActivityWidgetState();
+}
+
+class _RecentTokenActivityWidgetState extends State<RecentTokenActivityWidget> {
+  late final Stream<Map<String, IList<TxCommon>>> _groupedTxs;
+
+  @override
+  void initState() {
+    super.initState();
+    _groupedTxs =
+        sl<TransactionRepository>().watchGroupedByDate(widget.tokenAddress);
+  }
+
+  String _formatDate(String date) {
+    final parsedDate = DateTime.parse(date);
+    final now = DateTime.now();
+    final yesterday = now.subtract(const Duration(days: 1));
+
+    return (parsedDate.year == now.year &&
+            parsedDate.month == now.month &&
+            parsedDate.day == now.day)
+        ? context.l10n.today
+        : (parsedDate.year == yesterday.year &&
+                parsedDate.month == yesterday.month &&
+                parsedDate.day == yesterday.day)
+            ? context.l10n.yesterday
+            : DateFormat('MMM d, yyyy').format(parsedDate);
+  }
+
+  @override
+  Widget build(BuildContext context) =>
+      StreamBuilder<Map<String, IList<TxCommon>>>(
+        stream: _groupedTxs,
+        builder: (context, snapshot) {
+          final data = snapshot.data;
+          if (snapshot.connectionState == ConnectionState.waiting) {
+            return const CircularProgressIndicator();
+          }
+
+          if (data == null || data.isEmpty) {
+            return const Center(child: _NoActivity());
+          }
+
+          final sortedDates = data.keys.toList()
+            ..sort((a, b) => b.compareTo(a));
+
+          return HomeTile(
+            padding: const EdgeInsets.symmetric(horizontal: 22, vertical: 32),
+            child: Column(
+              crossAxisAlignment: CrossAxisAlignment.start,
+              mainAxisSize: MainAxisSize.max,
+              children: [
+                const SizedBox(height: 16),
+                ...sortedDates.map((date) {
+                  final transactions = data[date];
+                  late final IList<TxCommon> sortedTxs;
```

Comment:
This looks error-prone. What if transactions == null? It will never be initialized.

Also, the whole mapping in widget looks too complex, let's extract it to logic layer – no bloc is needed but something lightweight like ChangeNotifier would make sense for preparing the data.